app/services/duplicate_canonical_profile_resolver.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+# Class that deduplicates canonical profiles with identical benchmark/ref IDs
+class DuplicateCanonicalProfileResolver
+  class << self
+    def run!
+      duplicates.each do |duplicate|
+        profiles = Profile.canonical.includes(:child_profiles).where(
+          duplicate.slice(:ref_id, :benchmark_id)
+        ).order(created_at: :asc).offset(1)
+
+        return duplicate_error_msg(duplicate) unless no_child_profiles? profiles
+
+        profiles.destroy_all
+        duplicate_success_msg(duplicate)
+      end
+    end
+
+    private
+
+    def duplicates
+      Profile.canonical.select(
+        :ref_id,
+        :benchmark_id,
+        :policy_id,
+        'COUNT(id)'
+      ).group(
+        :ref_id,
+        :benchmark_id,
+        :policy_id
+      ).having('count(id) > 1')
+    end
+
+    def no_child_profiles?(profiles)
+      profiles.all? { |p| p.child_profiles.count.zero? }
+    end
+
+    def duplicate_success_msg(duplicate)
+      Rails.logger.info(%(
+        Deleted #{duplicate.count - 1} duplicates of policy with
```

Comment:
`destroy_all` returns the records it destroyed. We usually rely on that to determine how many were deleted, but this would seem to match that count. It's only ever going to run once, so I'm not asking for a change here if @vkrizan is also OK with it.

Reply:
Using `duplicate` seems reasonable, considering the error message printed prior to destroy.

@skateman You might want to check whether general logging is printed out on a migration.

Reply:
Yeah, in the migration context there's `say` and `write` which isn't available from the context of the service. So I'm going with `warn` and `STDOUT.puts` instead as they migrations rely on these anyway and rubocop likes them too.